packages/in_app_purchase/lib/src/in_app_purchase_tizen_platform.dart

```diff
@@ -0,0 +1,185 @@
+    }
+
+    final ProductDetailsResponse productDetailsResponse =
+        ProductDetailsResponse(
+      productDetails: productDetailsList,
+      notFoundIDs: invalidMessage,
+      error: exception == null
+          ? null
+          : IAPError(
+              source: kIAPSource,
+              code: exception.code,
+              message: exception.message ?? '',
+              details: exception.details),
+    );
+    return productDetailsResponse;
+  }
+
+  @override
+  Future<void> restorePurchases({
+    String? applicationUserName,
+  }) async {
+    List<GetUserPurchaseListAPIResult> responses;
+
+    responses = await Future.wait(<Future<GetUserPurchaseListAPIResult>>[
+      billingManager.requestPurchases(
+          applicationUserName, _identifiers.toList())
+    ]);
+
+    final List<PurchaseDetails> pastPurchases =
+        responses.expand((GetUserPurchaseListAPIResult response) {
+      return response.invoiceDetails;
+    }).map((InvoiceDetails purchaseWrapper) {
+      final SamsungCheckoutPurchaseDetails purchaseDetails =
+          SamsungCheckoutPurchaseDetails.fromPurchase(purchaseWrapper);
+
+      purchaseDetails.status = PurchaseStatus.restored;
+      return purchaseDetails;
+    }).toList();
+    _purchaseUpdatedController.add(pastPurchases);
+  }
+
+  @override
+  Future<bool> buyNonConsumable({required PurchaseParam purchaseParam}) async {
+    final BillingBuyData billingResultWrapper = await billingManager.buyItem(
+        appId: _appId,
+        serverType: _serverType,
+        orderItemId: purchaseParam.productDetails.id,
+        orderTitle: purchaseParam.productDetails.title,
+        orderTotal: purchaseParam.productDetails.price,
+        orderCurrencyId: purchaseParam.productDetails.currencyCode);
+
+    if (billingResultWrapper.payResult == 'SUCCESS') {
+      final String invoiceId =
+          billingResultWrapper.payDetails['InvoiceID'] ?? '';
+      final GetUserPurchaseListAPIResult responses = await billingManager
+          .requestPurchases(_identifiers.toList()[8], _identifiers.toList());
+
+      for (int i = 0; i < responses.invoiceDetails.length; i++) {
+        if (responses.invoiceDetails[i].invoiceID != invoiceId) {
+          final List<PurchaseDetails> purchases = <PurchaseDetails>[];
+          purchases.add(PurchaseDetails(
+            purchaseID: responses.invoiceDetails[i].invoiceID,
+            productID: responses.invoiceDetails[i].itemID,
+            verificationData: PurchaseVerificationData(
+                localVerificationData: responses.invoiceDetails[i].invoiceID,
+                serverVerificationData: responses.invoiceDetails[i].invoiceID,
+                source: kIAPSource),
+            transactionDate: responses.invoiceDetails[i].orderTime,
+            status: const PurchaseStateConverter()
+                .toPurchaseStatus(responses.invoiceDetails[i].appliedStatus),
+          ));
+
+          _purchaseUpdatedController.add(purchases);
```

Comment:
`buyNonConsumable` is supposed to return immediately, regardless of the actual result of the purchase, according to the API description:

> This method does not return the result of the purchase. Instead, after triggering this method, purchase updates will be sent to [purchaseStream].

So I'd suggest replacing the `await billingManager.requestPurchases()` part with `Future.then()` as follows:

```dart
billingManager
    .requestPurchases(...)
    .then((GetUserPurchaseListAPIResult responses) {
  for (...) {

  }
});
```
In this way the `requestPurchases` body is executed asynchronously after `buyNonConsumable` returns. (Consider adding `catchError` as well if `requestPurchases` may throw.)

Reply:
fixed.
[Use Future.then() in buyNonConsumable](https://github.com/flutter-tizen/plugins/pull/562/commits/ac657875a0678afdb13bf91f5bb01df8ff1e8075)

Reply:
I said `catchError` (`Future.then(...).catchError(...)`) but not try-catch. The try-catch statement added to the commit seems pointless. You should `addError` to the stream controller (`_purchaseUpdatedController`) if you want to report any error.

Reply:
fixed.
[add catchError if requestPurchases throw](https://github.com/flutter-tizen/plugins/pull/562/commits/632ba7fdef9f8d3bc05f96c9a45a9c34870cbf25)

Reply:
You shouldn't throw anything in the `catchError` clause. No one can handle the exception.

As I mentioned above, you can either `addError` to the stream controller or just ignore the error.

Reply:
ok,  just ignore error